lib/collection/src/shards/shard_holder/reshardable_read_request.rs

```diff
@@ -0,0 +1,109 @@
+use std::sync::Arc;
+
+use segment::data_types::facets::FacetRequest;
+use segment::types::Filter;
+
+use crate::operations::types::{CoreSearchRequest, CoreSearchRequestBatch, CountRequestInternal};
+use crate::operations::universal_query::shard_query::ShardQueryRequest;
+use crate::shards::shard::ShardId;
+
+pub trait EditFilter {
+    fn edit_filter<E>(&mut self, edit: E)
+    where
+        E: Clone + FnMut(Option<Filter>) -> Option<Filter>;
+}
+
+// Preserves original request, and optionally holds an edited one with the resharding filter.
+//
+// The edited request should be used on all shards, except the new resharding shard
+pub struct ReshardableReadRequest<T> {
+    /// Original request
+    pub original: Arc<T>,
+
+    /// Tuple of edited request and resharding shard id
+    pub edited: Option<(Arc<T>, u32)>,
+}
+
+impl<T: EditFilter + Clone> ReshardableReadRequest<T> {
+    pub fn new(request: Arc<T>, resharding_id_and_filter: Option<(u32, Filter)>) -> Self {
+        let Some((resharding_id, resharding_filter)) = resharding_id_and_filter else {
+            return Self {
+                original: request,
+                edited: None,
+            };
+        };
+
+        let mut edited = request.as_ref().clone();
+
+        edited.edit_filter(|filter| {
+            Some(match filter {
+                Some(filter) => filter.merge_owned(resharding_filter.clone()),
+                None => resharding_filter.clone(),
+            })
+        });
+
+        Self {
+            original: request,
+            edited: Some((Arc::new(edited), resharding_id)),
+        }
+    }
+
+    /// If available, use edited request on all shards, except the new resharding shard
+    pub fn get_request(&self, shard_id: ShardId) -> Arc<T> {
+        if let Some((edited, resharding_id)) = self.edited.clone() {
+            // Use edited request on all shards, except the new resharding shard
+            if resharding_id != shard_id {
+                return edited;
+            }
+        }
```

Comment:
We can probably eliminate this clone, by checking the shard ID as a reference first.

Reply:
Done as part of my changes.